### PR TITLE
Proper Testing & better use of Lux layers

### PR DESCRIPTION
**Pull request by avik-pal**
Supercedes #56  and #54

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Can some of the adjoints here be dropped now that NonlinearSolve.jl directly supports the adjoints?

(@frankschae )

**Comment by avik-pal** [comment]
The only difference from the upstreamed version is `jacobian free backprop` which is a trivial check, and the other one is it works with ps as NamedTuple using fmap. Would it be fine to add a dep on Functors?

(Initially, it made sense to have this here so that I could iterate faster, but rn there is no other blocker for upstreaming)

**Comment by ChrisRackauckas** [comment]
That could be done. We've been discussing canonicalization here: https://github.com/SciML/DifferentialEquations.jl/issues/881

File: src/adjoint.jl [hunk 1]
```diff
@@ -1,84 +1,122 @@
...
+import UnPack
+import Zygote
 
-@noinline function SciMLSensitivity.SteadyStateAdjointProblem(sol::EquilibriumSolution,
-                                                              sensealg::DeepEquilibriumAdjoint,
-                                                              g::Nothing, dg;
-                                                              save_idxs=nothing, kwargs...)
-  @unpack f, p, u0 = sol.prob
+# TODO(@avik-pal): Move to Lux.jl or maybe CRC?
+function CRC.rrule(::Type{T}, args...) where {T <: NamedTuple}
```
Comment (ChrisRackauckas) [inline src/adjoint.jl:11]: CRC I would think?

**Comment by ChrisRackauckas** [comment]
Yeah I think it's fine to merge and try upstreaming some of those pieces in pieces.

**Pull request merged**